



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.">
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.5, mkdocs-material-2.9.2">
    
    
      
        <title>TICR pipeline - PhyloNetworks.jl</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application.ba0fd1a6.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-palette.6079476c.css">
      
      
        
        
        <meta name="theme-color" content="#009688">
      
    
    
      <script src="../../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="teal" data-md-color-accent="teal">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#ticr-pipeline" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="PhyloNetworks.jl" class="md-header-nav__button md-logo">
          
            <img src="../../snaq_small.png" width="24" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                PhyloNetworks.jl
              </span>
              <span class="md-header-nav__topic">
                TICR pipeline
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <span class="md-nav__button md-logo">
      
        <img src="../../snaq_small.png" width="48" height="48">
      
    </span>
    PhyloNetworks.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../inputdata/" title="Input Data for SNaQ" class="md-nav__link">
      Input Data for SNaQ
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        TICR pipeline
      </label>
    
    <a href="./" title="TICR pipeline" class="md-nav__link md-nav__link--active">
      TICR pipeline
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#to-run-mrbayes-we-already-have-alignments" title="To run MrBayes: we already have alignments" class="md-nav__link">
    To run MrBayes: we already have alignments
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#no-scheduler-pipeline" title="no-scheduler pipeline" class="md-nav__link">
    no-scheduler pipeline
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#slurm-pipeline" title="slurm pipeline" class="md-nav__link">
    slurm pipeline
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../snaq_plot/" title="Network estimation and display" class="md-nav__link">
      Network estimation and display
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dist_reroot/" title="Network comparison and manipulation" class="md-nav__link">
      Network comparison and manipulation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../fixednetworkoptim/" title="Candidate Networks" class="md-nav__link">
      Candidate Networks
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../expectedCFs/" title="Extract Expected CFs" class="md-nav__link">
      Extract Expected CFs
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../bootstrap/" title="Bootstrap" class="md-nav__link">
      Bootstrap
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../multiplealleles/" title="Multiple Alleles" class="md-nav__link">
      Multiple Alleles
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trait_tree/" title="Continuous Trait Evolution" class="md-nav__link">
      Continuous Trait Evolution
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Library
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Library
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/public/" title="Public" class="md-nav__link">
      Public
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/internals/" title="Internals" class="md-nav__link">
      Internals
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#to-run-mrbayes-we-already-have-alignments" title="To run MrBayes: we already have alignments" class="md-nav__link">
    To run MrBayes: we already have alignments
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#no-scheduler-pipeline" title="no-scheduler pipeline" class="md-nav__link">
    no-scheduler pipeline
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#slurm-pipeline" title="slurm pipeline" class="md-nav__link">
    slurm pipeline
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/crsl4/PhyloNetworks.jl/edit/master/docs/man/ticr_howtogetQuartetCFs.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='TICR-pipeline-1'></a></p>
<h1 id="ticr-pipeline">TICR pipeline</h1>
<p>PhyloNetworks' <a href="https://github.com/crsl4/PhyloNetworks.jl/wiki/TICR:-from-alignments-to-quartet-concordance-factors">wiki</a> has a step-by-step tutorial, to go from multiple sequence alignments to a table of quartet gene frequencies (concordance factors: CFs), through BUCKy (to integrate out gene tree uncertainty) or through RAxML. To get the <code>raxml.pl</code> perl script to run RAxML on each gene, download the content of that wiki with</p>
<p><code>git clone https://github.com/crsl4/PhyloNetworks.jl.wiki.git</code></p>
<p>then go to the <code>script/</code> folder.   Full information and code is <a href="https://github.com/nstenz/TICR">here</a>. Below is more detail to insert data into the pipeline at various stages, using one of <strong>two pipelines</strong>, depending on your machine configuration:</p>
<ul>
<li>"no-scheduler" pipeline: original set of Perl scripts from the TICR pipeline, well suited for a machine or a cluster of machines without a job scheduler. The scripts automatically parallelize the work across the available cores.</li>
<li>"slurm" pipeline: well suited for a cluster where users submit jobs via a job scheduler like <a href="https://slurm.schedmd.com/">SLURM</a> or <a href="https://en.wikipedia.org/wiki/Oracle_Grid_Engine">SGE</a>. The job scheduler does the work of parallelizing the work across available cores. The scripts, in this second pipeline, were created to take full advantage of job scheduler capabilities. They were developed for a cluster running SLURM. Adjustments to the submit scripts will be needed, to adapt to your own SLURM configuration or to the syntax that your job scheduler wants.</li>
</ul>
<p><a id='To-run-MrBayes:-we-already-have-alignments-1'></a></p>
<h2 id="to-run-mrbayes-we-already-have-alignments">To run MrBayes: we already have alignments</h2>
<p><a id='no-scheduler-pipeline-1'></a></p>
<h3 id="no-scheduler-pipeline">no-scheduler pipeline</h3>
<p>We don't need to run <code>mdl.pl</code> if we already have aligned gene sequences from separate loci. To run MrBayes on each locus, we can simply create a tarball of the Nexus files we wish to use (fasta won't work at this stage). The command below assumes that we want to use all the files ending with ".nex" in the current directory, one file per locus:</p>
<div class="codehilite"><pre><span></span>tar czf my-genes.tar.gz *.nex
</pre></div>


<p>Once the tarball has been successfully generated, we can then specify this file as input for <a href="https://github.com/nstenz/TICR/blob/master/scripts/mb.pl">mb.pl</a> assuming we have a valid MrBayes block located in the file "bayes.txt":</p>
<div class="codehilite"><pre><span></span>mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb
</pre></div>


<p>If we get an error message like <code>mb.pl: Command not found</code>, it might be because <code>mb.pl</code> has no execute permission, or the current directory is not in our "path". An easy fix is to run this command instead:</p>
<div class="codehilite"><pre><span></span>perl mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb
</pre></div>


<p>The resulting output tarball would now be located in <code>my-genes-mb/my-genes.mb.tar</code>, and can be used normally with <a href="https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl">bucky.pl</a>, that is, like this:</p>
<div class="codehilite"><pre><span></span>bucky.pl my-genes-mb/my-genes.mb.tar -o mygenes-bucky
</pre></div>


<p>The output, with the table of concordance factors for all sets of 4 taxa, will be in a file named <code>my-genes.CFs.csv</code> inside a directory named <code>mygenes-bucky</code>. That's the file containing the quartet concordance factors to give to SNaQ as input. There is <em>no</em> need to do any of the steps below: they are already done by <code>bucky.pl</code>.</p>
<p><a id='slurm-pipeline-1'></a></p>
<h3 id="slurm-pipeline">slurm pipeline</h3>
<p>SLURM will parallelize the MrBayes runs across genes.</p>
<ol>
<li>Place all nexus files in a given folder.</li>
</ol>
<p>Include in this folder a text file with the MrBayes block for all the genes.   If we want a different MrBayes block per gene, step 2 should be skipped,   and we should instead find some other way to put the specific MrBayes block   at the end of each nexus file.</p>
<ol>
<li>Copy the julia script</li>
</ol>
<p><a href="https://github.com/nstenz/TICR/blob/master/scripts-cluster/paste-mb-block.jl"><code>paste-mb-block.jl</code></a>   and run it inside this folder. This script will read all the   nexus files in the directory, will read the text file with the MrBayes block to paste   onto all nexus files (option <code>‑m, ‑‑mb‑block</code> in the original <code>mb.pl</code> script),   will create a new directory, and will create new nexus files (containing the MrBayes block)   as <code>1.nex, 2.nex, ...</code> in the new directory. A <code>translate.txt</code> file will also be created   to map the original gene file names to the new (numbered) file names.</p>
<ol>
<li>Modify the submit script</li>
</ol>
<p><a href="https://github.com/nstenz/TICR/blob/master/scripts-cluster/mb-slurm-submit.sh"><code>mb-slurm-submit.sh</code></a>,   which will parallelize all the individual-gene MrBayes runs with SLURM:</p>
<ul>
<li>change <code>--array</code> to the correct number of genes</li>
<li>change <code>--mail-user</code> to the user's email (if this is an option for your job scheduler)</li>
<li>replace the <code>/workspace/software/bin</code> in <code>PATH="/workspace/software/bin:$PATH"</code> to the path where the <code>mb</code> executable is located or put the whole path in the command: <code>/s/mrbayes-3.2.6-1/bin/mb</code></li>
</ul>
<p>In slurm, we can then submit the MrBayes array job with:</p>
<div class="codehilite"><pre><span></span>  sbatch mb-slurm-submit.sh
  <span class="sb">```</span>

  With this slurm pipeline, the steps below need to be <span class="k">done</span>: keep reading.

<span class="c1">## To run mbsum on the output of MrBayes for each gene</span>

If we have the output of MrBayes and want to run BUCKy,
we must first run <span class="sb">`</span>mbsum<span class="sb">`</span> on the output from MrBayes, separately <span class="k">for</span> each gene.

For a gene with output tree files named <span class="sb">`</span>gene1.run1.t<span class="sb">`</span>, <span class="sb">`</span>gene1.run2.t<span class="sb">`</span> and <span class="sb">`</span>gene1.run3.t<span class="sb">`</span>,
and a desired burnin of <span class="m">1000</span> trees per tree file, we <span class="k">do</span> this:
</pre></div>


<p>bash mbsum -n 1000 -o gene1.in gene1.run1.t gene1.run2.t gene1.run3.t</p>
<div class="codehilite"><pre><span></span>This `mbsum` command will need to be executed for each gene.
Then we can continue to the next section to run bucky.

Alternatively, we can use the julia script
[`mbsum-t-files.jl`](https://github.com/nstenz/TICR/blob/master/scripts-cluster/mbsum-t-files.jl),
from within the directory that has the output tree files from MrBayes,
to run mbsum for *all* the genes.
`mbsum` is fast, so there is no attempt to parallelize the various mbsum commands.
</pre></div>


<p>bash julia mbsum-t-files.jl</p>
<div class="codehilite"><pre><span></span>**Warning:** this script is hard-coded to 3 independent runs per gene in MrBayes, and
for 2500 generations of burnin. Both can easily be changed:
edit this very short script and adapt to your number of runs and burnin.

## To run bucky on all 4-taxon sets: we already have the mbsum output

### no-scheduler pipeline

If we already ran `mbsum` on the output from MrBayes, for each individual gene,
we can simply create a tarball containing all the mbsum output files.
So if we had mbsum output in files named `gene1.in`, `gene2.in`, ... , `gene100.in`,
we would want to run something similar to the following command to create the tarball:
</pre></div>


<p>bash tar czf my-genes-mbsum.tar.gz gene*.in</p>
<div class="codehilite"><pre><span></span>We can now use this tarball along with the `-s` option in
[bucky.pl](https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl) like this:
</pre></div>


<p>bash bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</p>
<div class="codehilite"><pre><span></span>Again, if we get an error like `bucky.pl: Command not found`, we could run instead
</pre></div>


<p>bash perl bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</p>
<div class="codehilite"><pre><span></span>The output, with the table of concordance factors for all sets of 4 taxa,
will be in a file named `my-genes.CFs.csv` inside directory `mygenes-bucky`.
That&#39;s the file containing the quartet concordance factors to give to SNaQ as input.

### slurm pipeline

We want to run `bucky` on every 4-taxon set.
SLURM will parallelize these jobs with the submit script
[`bucky-slurm-submit.sh`](https://github.com/nstenz/TICR/blob/master/scripts-cluster/bucky-slurm-submit.sh),
which calls the perl script
[`bucky-slurm.pl`](https://github.com/nstenz/TICR/blob/master/scripts-cluster/bucky-slurm.pl).

The perl script
[`bucky-slurm.pl`](https://github.com/nstenz/TICR/blob/master/scripts-cluster/bucky-slurm.pl)
runs `bucky` on a single 4-taxon set.
It takes the following arguments, which must be modified in the submit script
[`bucky-slurm-submit.sh`](https://github.com/nstenz/TICR/blob/master/scripts-cluster/bucky-slurm-submit.sh):
- name of the folder containing the `mbsum` output files from previous step
  (this folder is named `mbsum` in the submit script: adapt)
- output name with option `-o`: option not used in the default submit script
- bucky arguments (with [options](https://github.com/nstenz/TICR#command-line-options-2), not used either)
- integer for the given quartet, via option `-q`.
  The quartet ID is specified by SLURM with its own array ID: `$SLURM_ARRAY_TASK_ID`.

In the submit script that gives instructions to the job scheduler:
- adapt the name of the `$SLURM_ARRAY_TASK_ID` variable, which captures the task number
  in the array of tasks, to your scheduler syntax
- change `--array` to the correct number of 4-taxon sets.
  For example, if there are 15 taxa in the dataset, there are `1365` 4-taxon sets.
  To get this number, if you are unsure, use `choose(15,4)` in R or `binomial(15,4)` in Julia,
  but replace 15 by your actual number of individuals.
- change `--mail-user` to the user&#39;s email (if this is an option for your job scheduler)
- replace the `/workspace/software/bin` in `PATH=&quot;/workspace/software/bin:$PATH&quot;`
  by the path where the `bucky` executable is located.
  Also, replace `/workspace/claudia/software/TICR/scripts/` by the full path where the
  `bucky-slurm.pl` script is located.


In slurm, you would submit the BUCKy array job with:
</pre></div>


<p>bash sbatch bucky-slurm-submit.sh</p>
<div class="codehilite"><pre><span></span>At the end, the array job will produce
- a `.concordance` file for every 4-taxon set
- a `.cf` file with the parsed output for that same 4-taxon set,
  in the format needed for the final CF table.

The `.cf` files can be concatenated to produce the file containing
the quartet concordance factors across all 4-taxon sets, to give to SNaQ as input:
</pre></div>


<p>bash cat *.cf &gt; CFtable.csv</p>
<div class="codehilite"><pre><span></span>Alternatively, if the list of `.cf` files is not easily captured by `*.cf`
(because the list is too long for a bash command), the following julia script
can do the concatenation. Just copy-paste the commands below within a Julia session,
started from the directory that contains the `.cf` files:
</pre></div>


<p>julia files = String[] # empty vector of strings: will contain the .cf file names later for f in filter(x -&gt; endswith(x, ".cf"), readdir())     push!(files,f) end println("found :(length(files)) cf files") # to check how many .cf output files were found open("CFtable.csv","w") do f_out   for file in files     @show file # to see the .cf file name: comment this out if that's too much screen output     open(file) do f_in         line = readstring(f_in)         write(f_out, string(line,"\n"))     end # closes "file" safely   end end # closes "CFtable.csv" safely ```</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../inputdata/" title="Input Data for SNaQ" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Input Data for SNaQ
              </span>
            </div>
          </a>
        
        
          <a href="../snaq_plot/" title="Network estimation and display" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Network estimation and display
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application.a59e2a89.js"></script>
      
      <script>app.initialize({version:"0.17.5",url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>