<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Estimate a Network from Concatenated DNA Sequences using PhyLiNC · PhyloNetworks.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">PhyloNetworks.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../inputdata/">Input Data for SNaQ</a></li><li><a class="tocitem" href="../ticr_howtogetQuartetCFs/">TICR pipeline</a></li><li><a class="tocitem" href="../snaq_plot/">Network estimation and display</a></li><li><a class="tocitem" href="../dist_reroot/">Network comparison and manipulation</a></li><li><a class="tocitem" href="../fixednetworkoptim/">Candidate Networks</a></li><li><a class="tocitem" href="../expectedCFs/">Extract Expected CFs</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../multiplealleles/">Multiple Alleles</a></li><li><a class="tocitem" href="../trait_tree/">Continuous Trait Evolution</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../fitDiscrete/">Discrete Trait Evolution</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Estimate a Network from Concatenated DNA Sequences using PhyLiNC</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Estimate a Network from Concatenated DNA Sequences using PhyLiNC</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/crsl4/PhyloNetworks.jl/blob/master/docs/src/man/phyLiNC.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Estimate-a-Network-from-Concatenated-DNA-Sequences-using-PhyLiNC"><a class="docs-heading-anchor" href="#Estimate-a-Network-from-Concatenated-DNA-Sequences-using-PhyLiNC">Estimate a Network from Concatenated DNA Sequences using PhyLiNC</a><a id="Estimate-a-Network-from-Concatenated-DNA-Sequences-using-PhyLiNC-1"></a><a class="docs-heading-anchor-permalink" href="#Estimate-a-Network-from-Concatenated-DNA-Sequences-using-PhyLiNC" title="Permalink"></a></h1><p>Estimate a phylogenetic network from concatenated DNA data using maximum likelihood, ignoring incomplete lineage sorting. Starting at a given starting tree or network, PhyLiNC uses a hill-climbing algorithm to search for the best network structure that best fits the given DNA sequences. (phyLiNC: phylogenetic Likelihood Network from Concatenated data)</p><p>The hill-climbing search is made up of nearest neighbor interchange moves, root change moves, add hybrid, and remove hybrid proposals.</p><h2 id="Getting-Started"><a class="docs-heading-anchor" href="#Getting-Started">Getting Started</a><a id="Getting-Started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-Started" title="Permalink"></a></h2><p>To estimate a phylogenetic network with PhyLiNC, you&#39;ll need concatenated DNA sequence data and a starting network.</p><p>Data: The DNA sequences should be in fasta format.</p><p>Starting Network or Tree: The starting network can be built using a tree-building tool like IQTree or inferred using the analyst&#39;s knowledge of the field. It should be in Newick string format. It should then be read in using PhyloNetwork&#39;s ReadTopology() function.</p><p>Usage: The network is constrained to have <code>maxhybrid</code> reticulations at most, but can be of any level. The search starts at (or near) the network <code>net</code>, using a local hill-climbing search to optimize the topology (nearest-neighbor interchange moves, add hybridizations, and remove hybridizations). Also optimized are evolutionary rates, amount of rate variation across sites, branch lengths and inheritance γs. This search strategy is run <code>nruns</code> times, and the best of the <code>nruns</code> networks is returned.</p><p>Return a <a href="man/@ref"><code>StatisticalSubstitutionModel</code></a> object, say <code>obj</code>, which contains the estimated network in <code>obj.net</code>.</p><p>Required arguments:</p><ul><li><code>net</code>: a network or tree of type <code>HybridNetwork</code>, to serve as a starting point in the search for the best network. Newick strings can be converted to this format with <a href="../../lib/public/#PhyloNetworks.readTopology"><code>readTopology</code></a>.</li><li><code>fastafile</code>: file with the sequence data in FASTA format.</li><li><code>substitutionModel</code>: A symbol indicating which substitution model is used. Choose <code>:JC69</code> f <a href="../../lib/public/#PhyloNetworks.JC69"><code>JC69</code></a> for the Jukes-Cantor model or <code>:HKY85</code> for the Hasegawa, Kishino, Yano model <a href="../../lib/public/#PhyloNetworks.HKY85"><code>HKY85</code></a>.</li></ul><h2 id="Network-Estimation"><a class="docs-heading-anchor" href="#Network-Estimation">Network Estimation</a><a id="Network-Estimation-1"></a><a class="docs-heading-anchor-permalink" href="#Network-Estimation" title="Permalink"></a></h2><p>By default, only three arguments are required for estimating a network with phyLiNC: a starting tree, a file of concatenated DNA sequences, and a subsitution model.</p><ul><li><code>net</code>: a network or tree of type <code>HybridNetwork</code>, to serve as a starting point in the search for the best network. Newick strings can be converted to this format with <a href="../../lib/public/#PhyloNetworks.readTopology"><code>readTopology</code></a>.</li><li><code>fastafile</code>: file with the sequence data in FASTA format.</li><li><code>substitutionModel</code>: A symbol indicating which substitution model is used. Choose <code>:JC69</code> f <a href="../../lib/public/#PhyloNetworks.JC69"><code>JC69</code></a> for the Jukes-Cantor model or <code>:HKY85</code> for the Hasegawa, Kishino, Yano model <a href="../../lib/public/#PhyloNetworks.HKY85"><code>HKY85</code></a>.</li></ul><pre><code class="language-julia">net = phyLiNC!(startingtree, fastafile, :JC69)</code></pre><p>At the beginning of optimization, the screen will show this: TODO add</p><p>At the end, it will give a final topology, as well as a model of evolution over the network: TODO add</p><h2 id="Beyond-Defaults:-Customize-Optimization-with-Optional-Arguments"><a class="docs-heading-anchor" href="#Beyond-Defaults:-Customize-Optimization-with-Optional-Arguments">Beyond Defaults: Customize Optimization with Optional Arguments</a><a id="Beyond-Defaults:-Customize-Optimization-with-Optional-Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Beyond-Defaults:-Customize-Optimization-with-Optional-Arguments" title="Permalink"></a></h2><p>In many cases, you may want to move beyond these defaults for a more customized estimation.</p><p>PhyLiNC&#39;s optional arguments (default value in parenthesis):</p><ul><li>symbol for the model of rate variation across sites (<code>:noRV</code> for no rate variation): use <code>:G</code> or <code>:Gamma</code> for Gamma-distributed rates, <code>:I</code> or <code>:Inv</code> for a proportion of invariable sites, and <code>:GI</code> or <code>:GammaInv</code> for a combination (not recommended).</li><li>integer (4) for the number of categories to use in estimating evolutionary rates using a discretized gamma model. When allowing for rate variation, four categories is standard. With 1 category, no rate variation is assumed. See <a href="../../lib/public/#PhyloNetworks.RateVariationAcrossSites"><code>RateVariationAcrossSites</code></a>.</li></ul><p>Main optional keyword arguments (default value in parenthesis):</p><ul><li><code>speciesfile</code> (&quot;&quot;): path to a csv file with samples in rows and two columns: species (column 1), individual (column 2) Include this file to group individuals by species.</li><li><code>cladefile</code> (&quot;&quot;): path to a csv file containing two columns: clades and individuals used to create one or more clade topology constraints to meet during the search. (NOTE: clade contraints not yet implemented.)</li><li><code>filename</code> (&quot;phyLiNC&quot;): root name for the output files (<code>.out</code>, <code>.err</code>). If empty (&quot;&quot;), files are <em>not</em> created, progress log goes to the screen only (standard out).</li><li><code>maxhybrid</code> (1): maximum number of hybridizations allowed. <code>net</code> (starting network) must have <code>maxhybrid</code> or fewer reticulations.</li><li><code>nruns</code> (10): number of independent starting points for the search</li><li><code>no3cycle</code> (true): prevents 3-cycles, which are (almost) not identifiable</li><li><code>nohybridladder</code> (true): prevents hybrid ladder in network. If true, the input network must not have hybrid ladders. Warning: Setting this to true will avoid most hybrid ladders, but some can still occur in some cases when deleting hybrid edges. This might be replaced with an option to avoid all non-tree-child networks in the future.</li></ul><p>Optional arguments controlling the search:</p><ul><li><code>seed</code> (default 0 to get it from the clock): seed to replicate a given search</li><li><code>nreject</code> (75): maximum number of times that new topologies are proposed and rejected in a row. Lower values of <code>nreject</code> result in a less thorough but faster search. Controls when to stop proposing new network topologies.</li><li><code>probST</code> (0.5): probability to use <code>net</code> as the starting topology for each given run. If probST &lt; 1, the starting topology is k NNI moves away from <code>net</code>, where k is drawn from a geometric distribution: p (1-p)ᵏ, with success probability p = <code>probST</code>.</li><li><code>maxmoves</code> (100): maximum number of topology moves before branch lengths, hybrid γ values, evolutionary rates, and rate variation parameters are reestimated.</li><li><code>verbose</code> (true): set to false to turn off screen output</li><li><code>alphamin</code> (0.02): minimum value for shape parameter alpha in rate variation across sites model.</li><li><code>alphamax</code> (50.0): maximum value for shape parameter alpha in rate variation across sites model.</li></ul><p>The following optional arguments control when to stop the optimization of branch lengths and gamma values on each individual candidate network. Defaults in parentheses.</p><ul><li><code>ftolRel</code> (1e-6) and <code>ftolAbs</code> (1e-6): relative and absolute differences of the network score between the current and proposed parameters</li><li><code>xtolRel</code> (1e-5) and <code>xtolAbs</code> (1e-5): relative and absolute differences between the current and proposed parameters.</li></ul><p>Greater values will result in a less thorough but faster search. These parameters are used when evaluating candidate networks only. Regardless of these arguments, once a final topology is chosen, branch lenghts are optimized using stricter tolerances (1e-10, 1e-12, 1e-10, 1e-10) for better estimates.</p><h2 id="Estimating-in-Parallel"><a class="docs-heading-anchor" href="#Estimating-in-Parallel">Estimating in Parallel</a><a id="Estimating-in-Parallel-1"></a><a class="docs-heading-anchor-permalink" href="#Estimating-in-Parallel" title="Permalink"></a></h2><p>By default, phyLiNC estimates ten networks, then returns the network with the highest likelihood. To speed this process, these ten runs can be done in parallel. For example, if your machine has more than four processors (or cores), you can tell julia to add 4 worker processors by starting julia with <code>julia -p 4</code>. This will give Julia five processors: one main and four workers. Alternatively, you can start Julia as usual (<code>julia</code>), then adding processors with:</p><pre><code class="language-julia">using Distributed
addprocs(4)</code></pre><p>If you&#39;re unsure how many you&#39;ve added, youcan check how many processors Julia currently has using <code>nprocs()</code>.</p><pre><code class="language-julia">nworkers()</code></pre><p>After adding processors, whether with <code>-p</code> or <code>addprocs()</code>, be sure to reload any packages you might need using the Distributed package&#39;s macro <code>@everywhere</code>. This ensures that all processors will have access to the packages.</p><p>After adding these processors, PhyLiNC will automatically run any estimation in this session using all cores.</p><pre><code class="language-julia">@everywhere using PhyloNetworks
net = PhyloNetworks.phyLiNC!(startingtree, fastafile, :JC69)</code></pre><p>TODO show output?</p><p>To plot your estimated networks you can use the visualization package, PhyloPlots.</p><pre><code class="language-">using PhyloPlots
using RCall # hide
R&quot;name &lt;- function(x) file.path(&#39;..&#39;, &#39;assets&#39;, &#39;figures&#39;, x)&quot; # hide
R&quot;svg(name(&#39;snaqplot_net0_1.svg&#39;), width=4, height=3)&quot; # hide
R&quot;par&quot;(mar=[0,0,0,0]) # hide
plot(net0, :R);
R&quot;dev.off()&quot;; # hide
nothing # hide</code></pre><h2 id="Branch-Length-Estimation-and-Identifiability"><a class="docs-heading-anchor" href="#Branch-Length-Estimation-and-Identifiability">Branch Length Estimation and Identifiability</a><a id="Branch-Length-Estimation-and-Identifiability-1"></a><a class="docs-heading-anchor-permalink" href="#Branch-Length-Estimation-and-Identifiability" title="Permalink"></a></h2><p>The length of the edge below a reticulation is not identifiable. Therefore, phyLiNC estimates the canonical version of the network: with reticulations <strong>unzipped</strong>: edges below reticulations are set to 0, and hybrid edges (parental lineages) have estimated lengths that are increased accordingly.</p><p>If any branch lengths are missing in the input network, phyLiNC estimates the starting branch lengths using pairwise distances. Otherwise, it uses the input branch lengths as starting branch lengths, only unzipping all reticulations, as described above.</p><h2 id="Choosing-a-Substitution-Model"><a class="docs-heading-anchor" href="#Choosing-a-Substitution-Model">Choosing a Substitution Model</a><a id="Choosing-a-Substitution-Model-1"></a><a class="docs-heading-anchor-permalink" href="#Choosing-a-Substitution-Model" title="Permalink"></a></h2><p>PhyLiNC allows you to choose from two Markov Process-based Evolutionary Rate Substitution models and four options for adding rate variation. These can be combined for eight possible ways to model evolutionary rates.</p><h3 id="Markov-Model-Substitution-Models"><a class="docs-heading-anchor" href="#Markov-Model-Substitution-Models">Markov Model Substitution Models</a><a id="Markov-Model-Substitution-Models-1"></a><a class="docs-heading-anchor-permalink" href="#Markov-Model-Substitution-Models" title="Permalink"></a></h3><p>There are two possible nucleic acid substitution models to choose from: JC69 (1969) and HKY85 (Hasegawa et al. 1985). The relative version of the model is used.</p><p>JC69: Jukes Cantor (1969) nucleic acid substitution model has a single rate parameter. <code>rate</code> corresponds to the absolute diagonal elements, that is, the rate of change (to any of the other 2 states). Individual rates are <code>rate</code>/3. The transition matrix <a href="../../lib/public/#PhyloNetworks.Q"><code>Q</code></a> is normalized to an average of 1 transition per unit of time: in which case <code>rate</code> is set to 1.0.</p><p>HKY85: A nucleic acid substitution model based on Hasegawa et al. 1985 substitution model. <code>rate</code> should be a vector of 1 or 2 rates, and <code>pi</code> a vector of 4 probabilities summing to 1. If <code>relative</code> is false, the 2 rates represent the transition rate and the transversion rate, α and β. If <code>relative</code> is true (default), only the first rate is used and represents the transition/transversion ratio: κ=α/β. The rate transition matrix Q is normalized to have 1 change / unit of time on average, i.e. the absolute version of Q is divided by <code>2(piT*piC + piA*piG)α + 2(piY*piR)β</code>.</p><p>Withever of these models, you choose, PhyLiNC will optimize the rate paramters throughout network estimation, then show the transition rate matrix <code>Q</code> for the final network when optimization is complete. The model is stored in the object that phyLiNC returns, which we&#39;ve called <code>obj</code> here. To explore the model further, use the following commands.</p><pre><code class="language-julia">julia&gt; model = obj.model
HKY85 Substitution Model base frequencies: [0.30095987201706437, 0.19137448340221302, 0.18764164778029596, 0.3200239968004266]
relative rate version with transition/tranversion ratio kappa = 4.03502,
 scaled so that there is one substitution per unit time
rate matrix Q:
               A       C       G       T
       A       *  0.1320  0.5223  0.2207
       C  0.2076       *  0.1294  0.8907
       G  0.8377  0.1320       *  0.2207
       T  0.2076  0.5327  0.1294       *

julia&gt; nstates(model)
4
julia&gt; getlabels(obj.model)
4-element Array{BioSymbols.DNA,1}:
 DNA_A
 DNA_C
 DNA_G
 DNA_T</code></pre><h3 id="Rate-Variation-and-Invariable-Sites"><a class="docs-heading-anchor" href="#Rate-Variation-and-Invariable-Sites">Rate Variation and Invariable Sites</a><a id="Rate-Variation-and-Invariable-Sites-1"></a><a class="docs-heading-anchor-permalink" href="#Rate-Variation-and-Invariable-Sites" title="Permalink"></a></h3><p>Concatenated models are well-suited to data with high percentages of invariable sites (TODO add citation). To this end, we provide functionality to estimate a rate of invariant sites using the invariable-sites model (+I, Hasegawa, Kishino &amp; Yano 1985 J Mol Evol). To enable this, run PhyLiNC like this:</p><pre><code class="language-julia">net = phyLiNC!(startingtree, fastafile, :JC69, :Inv)</code></pre><p>We can also estimate a variable rates across sites using the discrete Gamma model (+G, Yang 1994, Journal of Molecular Evolution). To use this model using the :G symbol in your PhyLiNC call:</p><pre><code class="language-julia">net = phyLiNC!(startingtree, fastafile, :JC69, :G)</code></pre><p>You can combine the two types of rate variation (+G+I, Gu, Fu &amp; Li 1995, Mol Biol Evol) but this is discouraged (Jia, Lo &amp; Ho 2014 PLOS One). Using rate variation increases the number of parameters by one (+G or +I) or by two (+G+I). Because the mean of the desired distribution or rates is 1, we use a Gamma distribution with shape α and scale θ=1/α (rate β=α) if no invariable sites, or scale θ=1/(α(1-pinv)), that is rate β=α(1-pinv) with a proportion pinv of invariable sites.</p><p>The shape parameter is referred to as alpha here.</p><p>The Gamma distribution is discretized into <code>ncat</code> categories.</p><p>In each category, the category&#39;s rate multiplier is a normalized quantile of the gamma distribution.</p><p>The rate model is also returned by PhyLiNC. To explore it further, use the following commands. For a network estimation with rate variation across sites:</p><pre><code class="language-julia">juila&gt; ratemodel = obj.ratemodel
Rate variation across sites: discretized Gamma
alpha: 2.0
categories for Gamma discretization: 4
rates: [0.319, 0.683, 1.109, 1.889]

julia&gt; nparams(ratemodel)
1</code></pre><p>For an invariable sites model:</p><pre><code class="language-julia">juila&gt; ratemodel = obj.ratemodel
Rate variation across sites: +I (invariable sites)
pinv: 0.3
rates: [0.0, 1.429]

julia&gt; nparams(ratemodel)
1</code></pre><p>If the network has both types of rate variation, the output will look like this:</p><pre><code class="language-julia">juila&gt; ratemodel = obj.ratemodel
Rate variation across sites: discretized Gamma+I
pinv: 0.3
alpha: 2.0
categories for Gamma discretization: 4
rates: [0.0, 0.456, 0.976, 1.584, 2.698]
probabilities: [0.3, 0.175, 0.175, 0.175, 0.175]

julia&gt; nparams(ratemodel)
2</code></pre></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 7 October 2020 13:36">Wednesday 7 October 2020</span>. Using Julia version 1.5.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
